### ISSUE TITLE
Design choice for Bijections (inspired by Distrax)

### ISSUE THREAD

**Issue by amanjitsk** [danielward27/flowjax#207]
Hi,

Thanks for the amazing package. I was going through the source for the `AbstractBijection` class and I had some questions/suggestions for the design choice concerning these transformations.

1. Is it really necessary to construct the `AbstractBijection` class with `shape` and `cond_shape` attributes, instead of something like `in_dims` (see `event_ndims_in` and `event_ndims_out` in `distrax.Bijector` for a reference). I actually find distrax's approach here to be more hassle-free here as it is common for many bijectors to be defined in a "dimension-agnostic" fashion. For example, the exp/log is a bijection and works the same way independent of the specific shape of the input/output. I think that flowjax's approach here is overly restrictive and (unnecessarily) binds bijection instances to a fixed shape (instead of just a fixed number of dimensions - which is usually sufficient).
2.  I understand this change (if it is something that you are open to) will have a downstream effect on things like `_VectorizedBijection.vectorize` and inferring the shape of `AbstractTransformed` distributions - but these can be taken care of fairly easily (distrax for instance just infers the shape of the transformed distribution using `jax.eval_shape`).

Will you be open to considering these changes to the `AbstractBijection` class ? I think these would make it a lot more hassle-free to define bijections without a bound specific shape (but rather just something like `event_ndims_in` and `event_ndims_out` - strictly speaking, the `event_ndims_out` isn't even really needed unless we'd like to define bijections with potentially redundant parameterizations/coordinates) - but to be clear I am just advocating to do away with `shape` and `cond_shape` and replace them with `ndims` and `cond_ndims` instead.

What do you think?

**Comment by danielward27** [comment]
Thanks! Often in JAX we know the shapes ahead of time as use of variable shapes triggers a recompilation, so for me the stricter approach gives some nicer shape checking (ensures an error occurs, and gives  a good error message), compared to using ndim (and hoping the bijection errors with an incorrect shape?). Beyond shape checking and the areas you mentioned, the primary thing impacted would be combining bijections, particularly ``Stack`` and ``Concatenate``, which require an explicitly defined shape.

Out of interest, what is your use case such that providing ndim is preferable to a shape?

**Comment by amanjitsk** [comment]
Yeah so like I said, my primary use case is "dimension-agnostic" bijectors. For example, we can take the `exp` and `log` functions as an example - these element-wise bijectors would be implemented in a dimension-free way in JAX.

```python
forward = lambda x: jnp.exp(x)
inverse = lambda y: jnp.log(y)
```

Now of course, the only thing that the logdet Jacobian requires is the dimensionality of the input, rather than its specific shape. So one could use the same bijector insance for vectors of any size `d`, instead of having to unnecessarily being tied down to a specific `bijector.shape` - `bijector.ndims` is usually sufficient to compute the logdets and infer shapes in push-forward distributions. Of course, in this case, one could argue that this bijector is only defined for scalars (just like you have implementations of the `Sigmoid`, `Tanh` and `Softplus` bijectors) - but here is a more realistic example: Consider the [ALR](https://en.wikipedia.org/wiki/Compositional_data#Additive_log_ratio_transform) transformation, commonly used in logistic regression and compositional data. It is defined on vectors (so `ndims=`), and bijects between the unit simplex and the real space

```python
forward = lambda p: jnp.log(p / (1 - jnp.sum(p))) 
inverse = lambda x: ...
``` 

The main use case is that I can carry around the same bijector instance if we adopt `ndims` instead of `shape` that binds down the bijector in this case to a specific dimension - which is not needed, and just adds an extra layer of threading the specific dimension size. 

Interestingly, actually the simplex example also demonstrates a use case where it might be useful to allow bijections with differing input and out `ndims` (or the more specific `shape` attribute at it is now). 

As for the `Stack` and `Concatenate` bijections, if anything adopting the `ndims` and `cond_ndims` attributes (instead of the more specific `shape` and `cond_shape`) would make that implementation easier and more potentially more general (by handling broadcasting etc.). But I don't think it should be an issue - if we do away with `shape` and replace it with `ndims`.

This also brings up another "type" of bijection that would be convenient to have - the `Block` bijection - that promotes element-wise (scalar) bijections to higher-dimensional bijections (by simply applying the base bijection element-wise), and the logdet just corresponds to summing across the promoted dimensions - see Distrax for a reference. And even this feature, would be implementable easily with the `ndims` attribute. With the `shape` attribute, it again introduces unnecessarily specific shape information - that would require a new bijection instance for different sized inputs (along the `ndims`). 

One thing that I'd like to add is that of course when needed, one can still provide the specific shape information for individual bijections that require it (like the neural examples, or for a "non-parameteric" examples that might need it - for example, something that partitions the vector into multiple components etc.).

I am happy to contribute to help make any of these changes if it interests you and is convincing enough!

**Comment by danielward27** [comment]
I can understand that allowing varying axis sizes can in some cases be convenient, but the benefits seem quite marginal. E.g., does one of the below options not work for your use case?
```python
import jax
from flowjax.bijections import Exp, Vmap

scalar_exp = Exp()

for i in range(1, 4):
    arr = jnp.arange(i)
    print(Exp(arr.shape).transform_and_log_det(arr))   # With reduction (option 1)
    print(Vmap(scalar_exp, axis_size=arr.size).transform_and_log_det(arr))   # With reduction  (option 2)
    print(jax.vmap(scalar_exp.transform_and_log_det)(arr))  # Without reduction
```
Again, this sort of programming with varying shapes is often an antipattern in JAX anyway. Note, ``Vmap`` (possibly with multiple applications) is our equivalent to ``Block``, but requires a specific axis size to be specified. 

``Concatenate`` only really makes sense given bijections with specified shape:
```python
from flowjax.bijections import Concatenate
from flowjax.bijections import Exp, Sigmoid

Concatenate([Exp((3, )), Sigmoid((1,))]).shape  # (4, )
```
Sure, you could pass the shapes as an argument if you had size agnostic bijections, but given many of the bijections will be defined for a specific shape anyway, that seems like it would messy and confusing.

I think a reasonable option would be to support a placeholder value in the shape tuples, to denote broadcasting compatibility (e.g. ``(None, )`` could represent any vector) - then at least all the bijections still have a "shape". But I'm not sure I'm convinced - it would lead to a bit of if-else cluttering for handling the different cases, increasing the code complexity for pretty minimal gain. It also seems like something that would be easy to forget to support when implementing new functions/bijections/etc that rely on the shape of a bijection (e.g. ``Reshape``, ``Chain``, ``Concatenate``). For now I think I'd favour the simplicity of the current approach of using a tuple of ints. I'd be open to facilitating access to broadcastable versions of bijection methods (something like the currently private ``_vectorize``) method, if that helps your use case.

You are correct that we don't support different ``in_shape`` and ``out_shape`` currently. I could potentially see that being something added in the future if needed.

**Comment by amanjitsk** [comment]
I guess that makes sense; I was just hoping to have the distrax approach as it seemed that most of the bijections I was working with can be implemented as an antipattern. Thanks for entertaining my thoughts and ramblings!
